src-ghc/Pact/GasModel/GasModel.hs

```diff
@@ -0,0 +1,331 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Pact.GasModel.GasModel where
+
+import Control.Exception              (bracket)
+import Control.Monad                  (void)
+import Statistics.Types               (Estimate(..))
+import Data.List                      (foldl', sortOn)
+import GHC.Conc                       (numCapabilities)
+
+import qualified Data.ByteString.Lazy.Char8 as BSL8
+import qualified Criterion.Main             as C
+import qualified Criterion                  as C
+import qualified Criterion.Types            as C
+import qualified Data.HashMap.Strict        as HM
+import qualified Data.Text                  as T
+import qualified Data.Csv                   as Csv
+import qualified Options.Applicative        as O
+
+
+-- Internal exports
+--
+import Pact.Eval                  (eval)
+import Pact.GasModel.Utils
+import Pact.GasModel.GasTests
+import Pact.GasModel.Types
+import Pact.Types.Lang
+import Pact.Types.Runtime         hiding (GasPrice)
+
+
+data Option = Option
+  { _oFilter :: Maybe T.Text
+  , _oBenchOnly :: Bool
+  }
+  deriving (Eq,Show)
+
+options :: O.ParserInfo Option
+options = O.info (O.helper <*> parser)
+          (O.fullDesc <> O.header "Gas Model Benchmarks")
+  where
+    parser =
+      Option
+      <$> ((Just <$> O.strOption
+            (O.short 'f' <> O.long "filter" <> O.metavar "NATIVE" <>
+             O.help "Pipe-delimited list of natives to run")) O.<|>
+            pure Nothing)
+      <*> (O.flag False True
+           (O.short 'b' <> O.long "bench" <> O.help "Just bench"))
+
+optionToGasTests
+  :: Option
+  -> [(NativeDefName, GasUnitTests)]
+optionToGasTests opt = HM.toList $
+  case _oFilter opt of
+    Nothing -> unitTests
+    Just ts -> HM.filterWithKey matchingFuns unitTests
+      where tlist = NativeDefName <$> T.split (== '|') ts
+            matchingFuns k _ = k `elem` tlist
+
+
+type NanoSeconds = Double
+
+secToNs :: Double -> NanoSeconds
+secToNs dInSeconds =
+  dInSeconds * 1000000000
+
+
+newtype Average = Average Double
+instance Csv.ToField Average where
+  toField (Average avg) = Csv.toField avg
+
+average :: [Double] -> Average
+average nums = Average avg
+  where meanSum = foldl' (+) 0.0 nums
```

Comment:
This version makes two passes over the input (and isn't strict enough), better is

```
average :: [Double] -> Average
average nums =
    let (total, num) = foldl' f (0.0, 0 :: Int) nums
        f (!t, !n) x = (t + x, n + 1)
    in Average $! total / fromIntegral num
```

If you want you can generalize to `Num a, Foldable f => average :: f a -> Average`